XiEditor/StyleMap.swift

```diff
@@ -108,4 +122,25 @@ class StyleMap {
                 applyStyle(string: string, id: styleSpan.style, range: styleSpan.range)
         }
     }
+
+    func stringWidth(str: NSString, style id: Int) -> CGFloat? {
+        if id >= map.count { return nil }
+        guard let style = map[id] else { return nil }
+
+        return str.size(withAttributes: style.attributes).width
+    }
+}
+
+func closestMatch(of font: NSFont, traits: NSFontTraitMask, weight: Int) -> NSFont? {
+    let fm = NSFontManager.shared()
+    var weight = weight
+    let fromWeight = fm.weight(of: font)
+    let direction = fromWeight > weight ? 1 : -1
+    while weight != fromWeight {
+        if let f = fm.font(withFamily: font.familyName ?? font.fontName, traits: traits, weight: weight, size: font.pointSize) {
+            return f
+        }
+        weight += direction
+    }
+    return nil
```

Comment:
I did an implementation with `availableMembers` first, but ended up with a less understandable code. If you like, I can show of the `availableMembers` variant of this implementation? (I also did a `convertWeight` implementation,  but this API did some really strange things).